curifactory/dag.py

```diff
@@ -0,0 +1,323 @@
+# programatically construct and analyze it this way.
+
+
+from curifactory.record import MapArtifactRepresentation, Record
+
+
+class ExecutionNode:
+    """Represents a particular stage for a particular record - a node in the
+    overall experiment graph.
+
+    Args:
+        record (Record): The record in which this stage would execute.
+        stage_name (str): The name of the stage that would execute.
+    """
+
+    def __init__(self, record: Record, stage_name: str):
+        self.record = record
+        self.stage_name = stage_name
+        self.parent: ExecutionNode = None
+        """The parent is a node that depends on this one/uses its outputs.
+        (This means that the same execution node might appear in multiple
+        execution trees, if its output is used by more than one other stage)"""
+
+        self.dependencies: list[ExecutionNode] = []
+        """Dependencies are the 'subtree' - the other nodes/stages that create
+        the outputs that match this node's inputs."""
+
+    def chain_rep(self) -> tuple[int, str]:
+        """Return this node represented as a tuple of the record index and stage name."""
+        return (self.record.get_record_index(True), self.stage_name)
+
+    def __str__(self):
+        return self.string_rep()
+
+    def string_rep(self, level=0) -> str:
+        """Recursively collect and return this node's index and name and that of
+        its subtrees."""
+        string = "\n"
+        string += "  " * level
+        string += f"({self.record.get_record_index(True)}, {self.stage_name})"
+        for child in self.dependencies:
+            string += child.string_rep(level + 1)
+        return string
+
+
+class DAG:
+    """A DAG represents an entire mapped version of an experiment as a graph, where
+    the nodes are stages and the connections are the outputs of one stage mapped to
+    the associated inputs of another.
+
+    The DAG is constructed as the first step of an experiment run (provided it isn't
+    run with ``no_dag=True`` or ``--no-dag`` on the CLI), by setting the artifact
+    manager to a special ``map_mode``. The experiment code is all run but every stage
+    short circuits before execution and after collecting information about it
+    (the record it's part of, what outputs are cached, etc.) This information is then
+    used to determine which stages actually need to execute, working backwards from
+    the leaf stages. This differs from curifactory's base ``no_dag`` mode because the
+    need-to-execute for every stage is based primarily on whether any future stage
+    actually requires this one's outputs and has a need-to-execute (resulting in a
+    recursive check backwards from the leaf stages.)
+    """
+
+    def __init__(self):
+        self.records: list[Record] = []
+        self.artifacts: list[MapArtifactRepresentation] = []
+        """This should essentially be an equivalent copy of ArtifactManager.artifacts.
```

Comment:
It's not a literal copy, `ArtifactManager.artifacts` is a list of `ArtifactRepresentation`s, instead of `MapArtifactRepresentation`s. But, either way, neither representation class stores the object itself, so it wouldn't be an issue